packages/expo-apple-authentication/ios/EXAppleAuthentication/EXAppleAuthentication.m

```diff
@@ -0,0 +1,200 @@
+// Copyright 2018-present 650 Industries. All rights reserved.
+
+#import <EXAppleAuthentication/EXAppleAuthentication.h>
+#import <EXAppleAuthentication/EXAppleAuthenticationMappings.h>
+
+#import <UMCore/UMDefines.h>
+#import <UMCore/UMUtilities.h>
+
+@interface EXAppleAuthentication ()
+
+@property (nonatomic, strong) UMPromiseResolveBlock promiseResolve;
+@property (nonatomic, strong) UMPromiseRejectBlock promiseReject;
+
+@property (nonatomic, weak) UMModuleRegistry *moduleRegistry;
+@property (nonatomic, weak) id <UMEventEmitterService> eventEmitter;
+@property (nonatomic, assign) BOOL hasListeners;
+
+@end
+
+@implementation EXAppleAuthentication
+
+UM_EXPORT_MODULE(ExpoAppleAuthentication);
+
+- (dispatch_queue_t)methodQueue
+{
+    return dispatch_get_main_queue();
+}
+
+- (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
+{
+  if (_moduleRegistry) {
+    [self invalidate];
+  }
+  _moduleRegistry = moduleRegistry;
+  _eventEmitter = [moduleRegistry getModuleImplementingProtocol:@protocol(UMEventEmitterService)];
+}
+
+- (NSArray<NSString *> *)supportedEvents
+{
+  return @[@"Expo.appleIdCredentialRevoked"];
+}
+
+- (void)invalidate
+{
+  _eventEmitter = nil;
+}
+
++ (BOOL)requiresMainQueueSetup
+{
+  return NO;
+}
+
+- (void)startObserving
+{
+  if (@available(iOS 13.0, *)) {
+    _hasListeners = YES;
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(credentialRevoked:)
+                                                 name:ASAuthorizationAppleIDProviderCredentialRevokedNotification
+                                               object:nil];
+  }
+}
+
+- (void)stopObserving
+{
+  if (@available(iOS 13.0, *)) {
+    _hasListeners = NO;
+    [[NSNotificationCenter defaultCenter] removeObserver:self
+                                                    name:ASAuthorizationAppleIDProviderCredentialRevokedNotification
+                                                  object:nil];
+  }
+}
+
+- (void)credentialRevoked:(NSNotification *)notification
+{
+  if (!_hasListeners) {
+    return;
+  }
+  [_eventEmitter sendEventWithName:@"Expo.appleIdCredentialRevoked" body:@{@"type": @"revoke"}];
+}
+
+UM_EXPORT_METHOD_AS(isAvailableAsync,
+                    isAvailableAsync:(UMPromiseResolveBlock)resolve
+                            rejecter:(UMPromiseRejectBlock)reject)
+{
+  if (@available(iOS 13.0, *)) {
+    return resolve(@(YES));
+  }
+  resolve(@(NO));
+}
+
+UM_EXPORT_METHOD_AS(requestAsync,
+                    requestAsync:(NSDictionary *)options
+                        resolver:(UMPromiseResolveBlock)resolve
+                        rejecter:(UMPromiseRejectBlock)reject)
+{
+  if (@available(iOS 13.0, *)) {
```

Comment:
```suggestion
  if (!@available(iOS 13.0, *)) {
     reject(...);
     return;
  }
  
  // we've narrowed the state space of the program so that we know apple auth is available here
  // and don't need to indent the code
```

Reply:
this produces an Xcode warning that `@available does not guard availability here; use if (@available) instead` so I'm going to leave as is. see https://stackoverflow.com/questions/52251973/ios-available-does-not-guard-availability-here-use-if-available-instead/52262260#52262260